plugins/python/pymodule.cpp

```diff
@@ -0,0 +1,134 @@
+  PyObject* mod = PyImport_ImportModule(modname.c_str());
+  if (mod) {
+    PyObject* reg = PyObject_GetAttrString(mod, "PHLEX_EXPERIMENTAL_REGISTER_ALGORITHMS");
+    if (reg) {
+      PyObject* pym = wrap_module(&m);
+      PyObject* pyconfig = wrap_configuration(&config);
+      if (pym && pyconfig) {
+        PyObject* res = PyObject_CallFunctionObjArgs(reg, pym, pyconfig, nullptr);
+        Py_XDECREF(res);
+      }
+      Py_XDECREF(pyconfig);
+      Py_XDECREF(pym);
+      Py_DECREF(reg);
+    }
+    Py_DECREF(mod);
+  }
+
+  if (PyErr_Occurred())
+    throw_runtime_error_from_py_error(false /* check_error */);
+}
+
+#ifdef PHLEX_HAVE_NUMPY
+static void import_numpy(bool control_interpreter)
+{
+  static std::atomic<bool> numpy_imported{false};
+  if (!numpy_imported.exchange(true)) {
+    if (_import_array() < 0) {
+      PyErr_Print();
+      if (control_interpreter)
+        Py_Finalize();
+      throw std::runtime_error("build with numpy support, but numpy not importable");
+    }
+  }
+}
+#endif
+
+static bool initialize()
+{
+  if (Py_IsInitialized()) {
+#ifdef PHLEX_HAVE_NUMPY
+    import_numpy(false);
+#endif
+    return true;
+  }
+
+  // TODO: the Python library is already loaded (b/c it's linked with
+  // this module), but its symbols need to be exposed globally to Python
+  // extension modules such as ctypes, yet this module is loaded with
+  // private visibility only. The workaround here locates the library and
+  // reloads (the handle is leaked b/c there's no knowing when it needs
+  // to be offloaded).
+  void* addr = dlsym(RTLD_DEFAULT, "Py_IsInitialized");
+  if (addr) {
+    Dl_info info;
+    if (dladdr(addr, &info) == 0 || info.dli_fname == 0 || info.dli_fname[0] == 0) {
+      throw std::runtime_error("unable to determine linked libpython");
+    }
+    dlopen(info.dli_fname, RTLD_GLOBAL | RTLD_NOW);
+  }
+
+#if PY_VERSION_HEX < 0x03020000
+  PyEval_InitThreads();
+#endif
+#if PY_VERSION_HEX < 0x03080000
+  Py_Initialize();
+#else
+  PyConfig config;
+  PyConfig_InitPythonConfig(&config);
+  PyConfig_SetString(&config, &config.program_name, L"phlex");
+  Py_InitializeFromConfig(&config);
+#endif
+#if PY_VERSION_HEX >= 0x03020000
+#if PY_VERSION_HEX < 0x03090000
+  PyEval_InitThreads();
+#endif
+#endif
+  // try again to see if the interpreter is now initialized
+  if (!Py_IsInitialized())
+    throw std::runtime_error("Python can not be initialized");
+
+#if PY_VERSION_HEX < 0x03080000
+  // set the command line arguments on python's sys.argv
+  wchar_t* argv[] = {const_cast<wchar_t*>(L"phlex")};
+  PySys_SetArgv(sizeof(argv) / sizeof(argv[0]), argv);
+#endif
+
+  // add custom types
+  PyType_Ready(&PhlexConfig_Type);
+  PyType_Ready(&PhlexModule_Type);
+  PyType_Ready(&PhlexLifeline_Type);
+
+  // load numpy (see also above, if already initialized)
+#ifdef PHLEX_HAVE_NUMPY
+  import_numpy(true);
+#endif
+
+  // TODO: the GIL should first be released on the main thread and this seems
```

Comment:
If you use a static RAII type that releases the GIL on construction and re-acquires on destruction that should solve this, and automatically handle only releasing it once:
https://en.cppreference.com/w/cpp/language/storage_duration.html#Static_block_variables